ssh/protocol.go

```diff
@@ -0,0 +1,264 @@
+	conn.mu.Lock()
+}
+
+func (conn *PktlineConnection) Unlock() {
+	conn.mu.Unlock()
+}
+
+func (conn *PktlineConnection) Start() error {
+	conn.Lock()
+	defer conn.Unlock()
+	return conn.negotiateVersion()
+}
+
+func (conn *PktlineConnection) End() error {
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessage("quit", nil, true)
+	if err != nil {
+		return err
+	}
+	_, _, err = conn.ReadStatus(true)
+	conn.cmd.Wait()
+	return err
+}
+
+func (conn *PktlineConnection) readCapabilities() ([]string, error) {
+	return conn.pl.ReadPacketList()
+}
+
+func (conn *PktlineConnection) negotiateVersion() error {
+	pkts, err := conn.readCapabilities()
+	if err != nil {
+		return errors.NewProtocolError("Unable to negotiate version with remote side (unable to read capabilities)", err)
+	}
+	ok := false
+	for _, line := range pkts {
+		if line == "version=1" {
+			ok = true
+		}
+	}
+	if !ok {
+		return errors.NewProtocolError("Unable to negotiate version with remote side (missing version=1)", nil)
+	}
+	err = conn.SendMessage("version 1", nil, false)
+	if err != nil {
+		return errors.NewProtocolError("Unable to negotiate version with remote side (unable to send version)", err)
+	}
+	status, _, err := conn.ReadStatus(true)
+	if err != nil {
+		return errors.NewProtocolError("Unable to negotiate version with remote side (unable to read status)", err)
+	}
+	if status != 200 {
+		return errors.NewProtocolError(fmt.Sprintf("Unable to negotiate version with remote side (unexpected status %d)", status), nil)
+	}
+	return nil
+}
+
+func (conn *PktlineConnection) SendMessage(command string, args []string, delim bool) error {
+	err := conn.pl.WritePacketText(command)
+	if err != nil {
+		return err
+	}
+	if delim {
+		err = conn.pl.WriteDelim()
+		if err != nil {
+			return err
+		}
+	}
+	for _, arg := range args {
+		err = conn.pl.WritePacketText(arg)
+		if err != nil {
+			return err
+		}
+	}
+	return conn.pl.WriteFlush()
+}
+
+func (conn *PktlineConnection) SendMessageWithLines(command string, args []string, lines []string) error {
+	err := conn.pl.WritePacketText(command)
+	if err != nil {
+		return err
+	}
+	for _, arg := range args {
+		err = conn.pl.WritePacketText(arg)
+		if err != nil {
+			return err
+		}
+	}
+	err = conn.pl.WriteDelim()
+	if err != nil {
+		return err
+	}
+	for _, line := range lines {
+		err = conn.pl.WritePacketText(line)
+		if err != nil {
+			return err
+		}
+	}
+	return conn.pl.WriteFlush()
+}
+
+func (conn *PktlineConnection) SendMessageWithData(command string, args []string, data io.Reader) error {
+	err := conn.pl.WritePacketText(command)
+	if err != nil {
+		return err
+	}
+	for _, arg := range args {
+		err = conn.pl.WritePacketText(arg)
+		if err != nil {
+			return err
+		}
+	}
+	err = conn.pl.WriteDelim()
+	if err != nil {
+		return err
+	}
+	buf := make([]byte, 32768)
+	for {
+		n, err := data.Read(buf)
+		if n > 0 {
+			conn.pl.WritePacket(buf[0:n])
+		}
+		if err != nil {
+			break
+		}
```

Comment:
We should be handling the error while writing, which I'll fix.  But we handle `n > 0` first because the error can be `io.EOF`, and the documentation for `Read` indicates that we should always handle a nonzero number of bytes read before handling the error.

Reply:
Ah, gotcha, thanks!